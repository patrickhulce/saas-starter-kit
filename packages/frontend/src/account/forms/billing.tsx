import * as React from 'react'
import {IUser} from '../../../../shared/lib/typedefs'

export class BillingForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2>Billing</h2>,
    ]
  }
}
