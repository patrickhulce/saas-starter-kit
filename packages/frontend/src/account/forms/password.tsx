import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'

export class PasswordForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2 key="password">Password</h2>,
    ]
  }
}
