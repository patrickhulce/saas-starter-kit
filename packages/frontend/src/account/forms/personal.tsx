import Button from '@material-ui/core/Button'
import TextField from '@material-ui/core/TextField'
import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'
import {ErrorBar} from '../../components/error-bar/error-bar'
import {Form, IFormData} from '../../components/form'
import {LoadingBar} from '../../components/loading-bar/loading-bar'

async function updateAccount(user: IUser): Promise<void> {
  const response = await fetch(`/api/v1/users/${user.id}`, {
    method: 'PUT',
    body: JSON.stringify(user),
    headers: {'content-type': 'application/json'},
  })

  if (response.status !== 200) {
    // TODO: Provide more specific error message
    throw new Error('Error updating account')
  }
}

interface INamesFormState {
  isLoading?: boolean
  errorMessage?: string
}

class NamesForm extends Form<{user: IUser}, INamesFormState> {
  public state: INamesFormState = {}

  protected async _handleSubmit(data: IFormData): Promise<void> {
    const user: IUser = {...this.props.user, firstName: data.first, lastName: data.last}
    await updateAccount(user)
  }

  public render(): JSX.Element {
    return (
      <form onSubmit={this.handleSubmit}>
        <LoadingBar isLoading={this.state.isLoading} />
        <ErrorBar message={this.state.errorMessage} />
        <TextField
          id="first-name"
          name="first"
          label="First Name"
          defaultValue={this.props.user.firstName}
          required
        />
        <TextField
          id="last-name"
          name="last"
          label="Last Name"
          defaultValue={this.props.user.lastName}
          required
        />
        <Button variant="raised" color="primary" type="submit">
          Update
        </Button>
      </form>
    )
  }
}

export class PersonalForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element[] {
    return [
      <h2 key="header">Personal Information</h2>,
      <NamesForm key="names-form" user={this.props.user} />,
    ]
  }
}
