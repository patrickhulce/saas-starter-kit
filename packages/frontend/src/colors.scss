$primary-brand: #0771b8;
$accent-brand: #fcb827;

$darker-gray: #333;
$dark-gray: #444;
$medium-gray: #666;
$lightest-gray: #eee;

$dark-red: #721c24;
$light-pink: #f5c6c6;
$lightest-pink: #fff4f4;

$dark-green: #1c721c;
$light-green: #c6f5cd;
$lightest-green: #f5fff4;
