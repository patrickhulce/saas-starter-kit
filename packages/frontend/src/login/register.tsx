import * as styles from './login.scss'
import * as React from 'react'
import Tabs, {Tab} from 'material-ui/Tabs'
import TextField from 'material-ui/TextField/TextField'
import Button from 'material-ui/Button/Button'
import FormControlLabel from 'material-ui/Form/FormControlLabel'
import Checkbox from 'material-ui/Checkbox/Checkbox'
import FormHelperText from 'material-ui/Form/FormHelperText'
import {Form} from './form'
import {login} from './login'

async function createAccount(
  accountName: string,
  firstName: string,
  lastName: string,
  email: string,
  password: string,
): Promise<any> {
  const response = await fetch('/api/v1/accounts/register', {
    method: 'POST',
    body: JSON.stringify({
      account: {name: accountName},
      user: {firstName, lastName, email, password},
    }),
    headers: {'content-type': 'application/json'},
  })

  if (response.status !== 200) {
    // TODO: Provide more specific error message
    throw new Error('Error creating account')
  }

  await login(email, password)
}

export class RegisterForm extends Form {
  public async _handleSubmit(data: any): Promise<void> {
    if (data.cpassword !== data.password) {
      this.setState({errorMessage: 'Passwords did not match'})
      return
    }

    try {
      await createAccount(
        `${data.first} ${data.last}'s Account`,
        data.first,
        data.last,
        data.email,
        data.password,
      )
    } catch (err) {
      this.setState({errorMessage: 'Error creating account'})
    }
  }

  public render(): JSX.Element {
    const errMsg = this.state.errorMessage ? <div>{this.state.errorMessage}</div> : undefined
    return (
      <form name="register" onSubmit={this.handleSubmit}>
        {errMsg}
        <div>
          <TextField name="first" type="first" label="First Name" className="half-text" required />
          <TextField name="last" type="last" label="Last Name" className="half-text" required />
        </div>
        <TextField name="email" type="email" label="Email" required />
        <TextField name="password" type="password" label="Password" required />
        <TextField name="cpassword" type="password" label="Confirm Password" required />
        <FormControlLabel
          control={<Checkbox name="newsletter" value="newsletter" color="primary" />}
          label="Send me tips and updates (unsubscribe anytime)"
        />
        <FormHelperText className={styles.tos}>
          By registering, you agree to our <a href="TOS_URL">terms of service</a>
        </FormHelperText>
        <Button
          className="atmn-create-account-submit"
          variant="raised"
          color="primary"
          type="submit"
        >
          Register
        </Button>
      </form>
    )
  }
}
