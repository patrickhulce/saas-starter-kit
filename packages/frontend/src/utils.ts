import * as React from 'react'
import * as ReactDOM from 'react-dom'

import {IUser} from '../../shared/lib/typedefs'

import {getLoggedInUser, refreshLoggedInUser} from './services/user-service'

export async function redirectToLogin(): Promise<void> {
  window.location.href = '/login'
  return new Promise(resolve => setTimeout(resolve, 10000)) as Promise<void>
}

export async function redirectIfNotAuthorized(): Promise<void> {
  try {
    const localUser = await getLoggedInUser()
    if (!localUser) await redirectToLogin()
    await refreshLoggedInUser()
    const remoteUser = await getLoggedInUser()
    if (!remoteUser || localUser!.id !== remoteUser.id) await redirectToLogin()
  } catch (err) {
    await redirectToLogin()
  }
}

export async function findUserOnStartupOrBail(): Promise<IUser> {
  const localUser = await getLoggedInUser()
  if (!localUser) await redirectToLogin()
  setTimeout(redirectIfNotAuthorized, 0)
  return localUser!
}

export function createRenderFn(render: () => React.ReactElement<any>): () => void {
  return () => ReactDOM.render(render(), document.getElementById('app-root'))
}

export function HMR(nodeModule: any, handle: (nodeModule: any) => any): void {
  if (nodeModule.hot) {
    handle(nodeModule)
  }
}

export const testIds = {
  loadingBar: 'loading-bar',
  messageBar: 'message-bar',
  // Login IDs
  createAccountTab: 'create-account-tab',
  createAccountSubmit: 'create-account-submit',
  loginFormPrimaryBtn: 'login-form-primary-btn',
  loginFormSecondaryBtn: 'login-form-secondary-btn',
  loginForm: 'login-form',
  registerForm: 'register-form',
  profileNamesForm: 'profile-names-form',
}
