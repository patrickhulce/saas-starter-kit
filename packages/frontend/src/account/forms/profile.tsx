import * as React from 'react'

import {IUser} from '../../../../shared/lib/typedefs'
import {BasicTextField} from '../../components/basic-text-field'
import {Form, IFormData, IFormState} from '../../components/form/form'
import * as formStyles from '../../components/form/form.scss'
import {updateAccount} from '../../services/user-service'
import {testIds} from '../../utils'

class NamesForm extends Form<{user: IUser}> {
  public state: IFormState = {}
  public testId: string = testIds.profileNamesForm

  protected async _handleSubmit(data: IFormData): Promise<void> {
    await updateAccount(this.props.user.id, {firstName: data.firstName, lastName: data.lastName})
  }

  public renderInputUI(): JSX.Element {
    return (
      <React.Fragment>
        <BasicTextField name="firstName" defaultValue={this.props.user.firstName} />
        <BasicTextField name="lastName" defaultValue={this.props.user.lastName} />
      </React.Fragment>
    )
  }
}

export class ProfileForm extends React.Component<{user: IUser}> {
  public render(): JSX.Element {
    const formClasses = `${formStyles.columnarForm} ${formStyles.fixedWidth}`
    return (
      <React.Fragment>
        <h2>Personal Information</h2>
        <NamesForm user={this.props.user} className={formClasses} />
      </React.Fragment>
    )
  }
}
