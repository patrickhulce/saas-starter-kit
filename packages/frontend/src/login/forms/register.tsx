import Checkbox from '@material-ui/core/Checkbox/Checkbox'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import FormHelperText from '@material-ui/core/FormHelperText'
import * as React from 'react'

import conf from '../../../../shared/lib/conf'
import {BasicTextField} from '../../components/basic-text-field'
import {Form, IFormState, ISubmitOptions} from '../../components/form'
import {testIds} from '../../utils'
import * as styles from '../login.scss'

import {login} from './login'

async function createAccount(
  accountName: string,
  firstName: string,
  lastName: string,
  email: string,
  password: string,
): Promise<any> {
  const response = await fetch('/api/v1/accounts/register', {
    method: 'POST',
    body: JSON.stringify({
      account: {name: accountName},
      user: {firstName, lastName, email, password},
    }),
    headers: {'content-type': 'application/json'},
  })

  if (response.status !== 200) {
    // TODO: Provide more specific error message
    throw new Error('Error creating account')
  }

  await login(email, password)
}

export class RegisterForm extends Form {
  public state: IFormState = {}
  public testId: string = testIds.registerForm
  public submitUIOptions: ISubmitOptions = {label: 'Register', testId: testIds.createAccountSubmit}

  protected async _handleSubmit(data: any): Promise<void> {
    if (data.confirmPassword !== data.password) {
      throw new Error('Passwords did not match')
    }

    await createAccount(
      `${data.firstName} ${data.lastName}'s Account`,
      data.firstName,
      data.lastName,
      data.email,
      data.password,
    )
  }

  public renderInputUI(): JSX.Element {
    const newsletterLabel = (
      <span className={styles.newsletter}>
        Send me email tips and updates
        <small>(Roughly once a month, unsubscribe anytime)</small>
      </span>
    )

    return (
      <React.Fragment>
        <div>
          <BasicTextField name="firstName" className={styles.halfText} autoFocus />
          <BasicTextField name="lastName" className={styles.halfText} />
        </div>
        <BasicTextField name="email" type="email" />
        <BasicTextField name="password" type="password" />
        <BasicTextField name="confirmPassword" type="password" />
        <FormControlLabel
          control={<Checkbox name="newsletter" value="newsletter" color="primary" />}
          label={newsletterLabel}
        />
        <FormHelperText className={styles.tos}>
          By registering, you agree to our{' '}
          <a target="_blank" href={conf.termsOfServiceURL}>
            terms of service
          </a>
        </FormHelperText>
      </React.Fragment>
    )
  }
}
