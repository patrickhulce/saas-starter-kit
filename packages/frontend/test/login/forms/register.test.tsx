import * as React from 'react'
import {RenderResult, fireEvent, render, wait} from 'react-testing-library'

import {RegisterForm} from '../../../src/login/forms/register'
import {testIds} from '../../../src/utils'
import {createFetchPromise} from '../../utils'

describe('register/forms/register.tsx', () => {
  let fetchMock: jest.Mock

  function renderAndFill(): RenderResult {
    const form = render(<RegisterForm />)
    const firstName = form.getByLabelText(/First Name/) as HTMLInputElement
    const lastName = form.getByLabelText(/Last Name/) as HTMLInputElement
    const email = form.getByLabelText(/Email/) as HTMLInputElement
    const password = form.getByLabelText(/^Password/) as HTMLInputElement
    const confirmPassword = form.getByLabelText(/Confirm Password/) as HTMLInputElement
    firstName.value = 'Klay'
    lastName.value = 'Thompson'
    email.value = 'klay@example.com'
    password.value = 'rocko'
    confirmPassword.value = 'rocko'
    return form
  }

  beforeEach(() => {
    fetchMock = jest.fn()
    self.fetch = fetchMock
  })

  it('should render form', () => {
    const form = render(<RegisterForm />)
    expect(form.container).toMatchSnapshot()
  })

  it('should check passwords match', async () => {
    const {getByTestId, getByLabelText} = renderAndFill()
    const confirmPassword = getByLabelText(/Confirm Password/) as HTMLInputElement
    confirmPassword.value = 'other'

    fireEvent.submit(getByTestId(testIds.registerForm))
    await wait(() => getByTestId(testIds.messageBar))

    expect(getByTestId(testIds.messageBar).textContent).toMatchSnapshot()
  })

  it('should show a loading UI', async () => {
    const {getByTestId, queryByTestId} = renderAndFill()
    const fetchPromise = createFetchPromise()
    fetchMock.mockImplementation(fetchPromise.fn)

    fireEvent.submit(getByTestId(testIds.registerForm))
    await wait(() => getByTestId(testIds.loadingBar))

    expect(queryByTestId(testIds.messageBar)).toBeNull()

    fetchPromise.reject(new Error('short-circuit'))
  })

  it('should send register request to server', async () => {
    const {getByTestId} = renderAndFill()
    const fetchPromise = createFetchPromise()
    fetchMock.mockImplementation(fetchPromise.fn)

    fireEvent.submit(getByTestId(testIds.registerForm))
    await wait(() => getByTestId(testIds.loadingBar))

    expect(fetchMock).toHaveBeenCalled()
    expect(fetchMock.mock.calls[0]).toMatchSnapshot()
  })

  it('should show an error UI', async () => {
    const {getByTestId, queryByTestId} = renderAndFill()
    const fetchPromise = createFetchPromise({status: 400})
    fetchMock.mockImplementation(fetchPromise.fn)

    fireEvent.submit(getByTestId(testIds.registerForm))
    await wait(() => getByTestId(testIds.loadingBar))

    expect(queryByTestId(testIds.messageBar)).toBeNull()

    fetchPromise.resolve()
    await wait(() => getByTestId(testIds.messageBar))

    expect(queryByTestId(testIds.loadingBar)).toBeNull()
    expect(getByTestId(testIds.messageBar).textContent).toMatchSnapshot()
  })
})
