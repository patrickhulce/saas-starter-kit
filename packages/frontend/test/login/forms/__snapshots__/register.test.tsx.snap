// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register/forms/register.tsx should check passwords match 1`] = `"Passwords did not match"`;

exports[`register/forms/register.tsx should render form 1`] = `
<div>
  <form
    data-testid="register-form"
    name="register"
  >
    <div>
      <div
        class="MuiFormControl-root-1 halfText"
      >
        <label
          class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
          data-shrink="false"
          for="first-name-input"
        >
          First Name
          <span
            class="MuiFormLabel-asterisk-14"
          >
             *
          </span>
        </label>
        <div
          class="MuiInput-root-15 MuiInput-formControl-16 MuiInput-underline-19"
        >
          <input
            aria-invalid="false"
            aria-required="true"
            class="MuiInput-input-23"
            id="first-name-input"
            name="firstName"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="MuiFormControl-root-1 halfText"
      >
        <label
          class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
          data-shrink="false"
          for="last-name-input"
        >
          Last Name
          <span
            class="MuiFormLabel-asterisk-14"
          >
             *
          </span>
        </label>
        <div
          class="MuiInput-root-15 MuiInput-formControl-16 MuiInput-underline-19"
        >
          <input
            aria-invalid="false"
            aria-required="true"
            class="MuiInput-input-23"
            id="last-name-input"
            name="lastName"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root-1"
    >
      <label
        class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink="false"
        for="email-input"
      >
        Email
        <span
          class="MuiFormLabel-asterisk-14"
        >
           *
        </span>
      </label>
      <div
        class="MuiInput-root-15 MuiInput-formControl-16 MuiInput-underline-19"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="MuiInput-input-23 MuiInput-inputType-26"
          id="email-input"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-1"
    >
      <label
        class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink="false"
        for="password-input"
      >
        Password
        <span
          class="MuiFormLabel-asterisk-14"
        >
           *
        </span>
      </label>
      <div
        class="MuiInput-root-15 MuiInput-formControl-16 MuiInput-underline-19"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="MuiInput-input-23 MuiInput-inputType-26"
          id="password-input"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="MuiFormControl-root-1"
    >
      <label
        class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
        data-shrink="false"
        for="confirm-password-input"
      >
        Confirm Password
        <span
          class="MuiFormLabel-asterisk-14"
        >
           *
        </span>
      </label>
      <div
        class="MuiInput-root-15 MuiInput-formControl-16 MuiInput-underline-19"
      >
        <input
          aria-invalid="false"
          aria-required="true"
          class="MuiInput-input-23 MuiInput-inputType-26"
          id="confirm-password-input"
          name="confirmPassword"
          required=""
          type="password"
          value=""
        />
      </div>
    </div>
    <label
      class="MuiFormControlLabel-root-28"
    >
      <span
        class="MuiButtonBase-root-46 MuiIconButton-root-40 MuiSwitchBase-root-36 MuiCheckbox-root-31 MuiCheckbox-colorPrimary-34"
      >
        <span
          class="MuiIconButton-label-45"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-49"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <input
            class="MuiSwitchBase-input-39"
            name="newsletter"
            type="checkbox"
            value="newsletter"
          />
        </span>
        <span
          class="MuiTouchRipple-root-55"
        />
      </span>
      <span
        class="MuiTypography-root-62 MuiTypography-body1-71 MuiFormControlLabel-label-30"
      >
        <span
          class="newsletter"
        >
          Send me email tips and updates
          <small>
            (Roughly once a month, unsubscribe anytime)
          </small>
        </span>
      </span>
    </label>
    <p
      class="MuiFormHelperText-root-86 tos"
    >
      By registering, you agree to our
       
      <a
        href="http://localhost/tos"
        target="_blank"
      >
        terms of service
      </a>
    </p>
    <button
      class="MuiButtonBase-root-46 MuiButton-root-90 MuiButton-contained-99 MuiButton-containedPrimary-100 MuiButton-raised-102 MuiButton-raisedPrimary-103"
      data-testid="create-account-submit"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label-91"
      >
        Register
      </span>
      <span
        class="MuiTouchRipple-root-55"
      />
    </button>
  </form>
</div>
`;

exports[`register/forms/register.tsx should send register request to server 1`] = `
Array [
  "/api/v1/accounts/register",
  Object {
    "body": "{\\"account\\":{\\"name\\":\\"Klay Thompson's Account\\"},\\"user\\":{\\"firstName\\":\\"Klay\\",\\"lastName\\":\\"Thompson\\",\\"email\\":\\"klay@example.com\\",\\"password\\":\\"rocko\\"}}",
    "headers": Object {
      "content-type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`register/forms/register.tsx should show an error UI 1`] = `"Error creating account"`;
